src/lib/statistics/statistics_objects/abstract_histogram.cpp

```diff
@@ -0,0 +1,1035 @@
+  }
+}
+
+template <typename T>
+Cardinality AbstractHistogram<T>::_invert_estimate(const Cardinality& estimate) const {
+  return Cardinality{total_count() - estimate};
+}
+
+template <typename T>
+std::shared_ptr<AbstractStatisticsObject> AbstractHistogram<T>::sliced(
+    const PredicateCondition predicate_condition, const AllTypeVariant& variant_value,
+    const std::optional<AllTypeVariant>& variant_value2) const {
+  if (_does_not_contain(predicate_condition, variant_value, variant_value2)) {
+    return nullptr;
+  }
+
+  const auto value = static_variant_cast<T>(variant_value);
+  if (!value) {
+    return clone();
+  }
+
+  switch (predicate_condition) {
+    case PredicateCondition::Equals: {
+      GenericHistogramBuilder<T> builder{1, _domain};
+      builder.add_bin(*value, *value,
+                      static_cast<HistogramCountType>(estimate_cardinality(PredicateCondition::Equals, variant_value)),
+                      1);
+      return builder.build();
+    }
+
+    case PredicateCondition::NotEquals: {
+      const auto value_bin_id = _bin_for_value(*value);
+      if (value_bin_id == INVALID_BIN_ID) return clone();
+
+      auto minimum = bin_minimum(value_bin_id);
+      auto maximum = bin_maximum(value_bin_id);
+      const auto distinct_count = bin_distinct_count(value_bin_id);
+
+      // Do not create empty bin, if `value` is the only value in the bin
+      const auto new_bin_count = minimum == maximum ? bin_count() - 1 : bin_count();
+
+      GenericHistogramBuilder<T> builder{new_bin_count, _domain};
+
+      builder.add_copied_bins(*this, BinID{0}, value_bin_id);
+
+      // Do not create empty bin, if `value` is the only value in the bin
+      if (minimum != maximum) {
+        // A bin [50, 60] sliced with `!= 60` becomes [50, 59]
+        // TODO(anybody) Implement bin bounds trimming for strings
+        if constexpr (!std::is_same_v<
+                          pmr_string,
+                          T>) {  // NOLINT clang-tidy is crazy and sees a "potentially unintended semicolon" here...
+          if (minimum == *value) {
+            minimum = _domain.next_value(*value);
+          }
+
+          if (maximum == *value) {
+            maximum = _domain.previous_value(*value);
+          }
+        }
+
+        const auto estimate = estimate_cardinality_and_distinct_count(PredicateCondition::Equals, variant_value);
+        const auto new_height = bin_height(value_bin_id) - estimate.cardinality;
+        const auto new_distinct_count = distinct_count - estimate.distinct_count;
+
+        builder.add_bin(minimum, maximum, new_height, new_distinct_count);
+      }
+
+      builder.add_copied_bins(*this, value_bin_id + 1, bin_count());
+
+      return builder.build();
+    }
+
+    case PredicateCondition::LessThanEquals:
+      return sliced(PredicateCondition::LessThan, _domain.next_value(*value));
+
+    case PredicateCondition::LessThan: {
+      auto last_bin_id = _bin_for_value(*value);
```

Comment:
This could be mistaken for the last bin of the histogram. `last_included_bin_id`?